src/main/java/org/rumbledb/runtime/flwor/clauses/GroupByClauseSparkIterator.java

```diff
@@ -165,25 +164,26 @@ private void setAllLocalResults() {
                     groupVariableExpression.close();
 
                     // if a new variable is declared inside the group by clause, insert value in tuple
-                    inputTuple.putValue(expression.getVariableReference().getVariableName(), newVariableResults);
+                    inputTuple.putValue(expression.getVariableName(), newVariableResults);
                     results.addAll(newVariableResults);
 
                 } else { // if grouping on a variable reference
-                    VariableReferenceIterator groupVariableReference = expression.getVariableReference();
```

Comment:
It's great to see that you've managed to tame this beast :) I had tried about a month ago but there was no way I could unentangle these as quickly as I had hoped for.